### ISSUE TITLE
Save terraform outputs to `guardian/outputs` branch

### ISSUE THREAD

**Issue by dcreey** [abcxyz/guardian#184]
### TL;DR

Guardian apply should optionally store outputs to a locked-down branch named `guardian/outputs`.

### Detailed design

```markdown
Why store outputs:
* Far more convenient to reference/link to (can link to a particular line)
* Don't have to remember which PR last applied changes to a particular entrypoint (the most up to date outputs are always in the branch)
* Stored outputs can be used to implement new features (abc generate templates using outputs as inputs)

Implementation:
* After a successful terraform apply the changed outputs should be committed to the `guardian/outputs` branch. ([terraform outputs](https://developer.hashicorp.com/terraform/cli/commands/output))
* the path to the outputs file should match the path to the tf entrypoint
* The commit should be pushed directly to the branch.
* The `guardian/outputs` branch should have branch protection to ensure no one has rights to commit to it besides the GitHub bot.
* 

Risks
* Technically anyone could create a github action in a PR and use it to commit/modify the outputs branch. Mitigation: It should be documented that this should be used for non-production/non-critical paths. Additionally, any reference to outputs in code should be locked by hash to prevent changes to head breaking things.
* Terraform outputs can store sensitive details. Mitigation: This should be documented and any output that is sensitive should be marked as sensitive. We should also not specify `-json` or `-raw` so that terraform will not ignore `sensitive` attributes.
```

### Alternatives considered

```markdown
1. Store the outputs in a GCS bucket. Why not: worse user experience
2. Store the outputs in the main branch. Why not: CLA blocks on github bot committing the code
```

**Comment by sethvargo** [comment]
Outputs are complex data structures (e.g. arrays of maps of objects of arrays of strings and numbers) - representing them as a file will be difficult unless you use `-json` (`-raw` does not support complex types). But as you noted, that will include sensitive values too. I also think it would be a mistake to assume end-users will correctly mark outputs as "sensitive" as a control mechanism.

Using a single branch assumes there is one active Guardian instance/actuation per repo, which I think is already untrue. If you name files after specific actuation stacks as suggested, users need to know _which _ file to read, and they have access to all of the other outputs.

Teams might have multiple actuations on different sub-folders, which would then clobber. The architecture you describe also has no mechanism for deleting stale output files for a stack that's deleted.

What are the use cases for doing this that can't be satisfied with Terraform remote state?

It sounds like there's a goal to have some database of values that are populated from Terraform. It would be better if those were hand-selected instead of assuming all outputs need to be published. Furthermore, publishing the outputs into a branch on a repo introduces authentication challenges (for the CLI use case you mentioned).

I don't think this is actually going to be as good of a user experience as you think.

**Comment by dcreey** [comment]
> Outputs are complex data structures (e.g. arrays of maps of objects of arrays of strings and numbers)
Anecdotally I did this with a pretty complex output for X's org and it seemed pretty readable. I can imagine it would get pretty crazy but maybe putting crazy values in outputs isn't actually helpful anyways?

> I also think it would be a mistake to assume end-users will correctly mark outputs as "sensitive" as a control mechanism.
These values are already all printed out in the PR comment. Anyone with access to the repo can go and see them. Doesn't seem like we are introducing any more risk. Additionally if it is that much of a risk perhaps we should run some kind of secret detection against the output and mask anything that seems like a secret (for PR comments too). If we did that then we could use `-json` or `-raw`.

> Using a single branch assumes there is one active Guardian instance/actuation per repo
I didn't elaborate on this - but my thought was that we would support any arbitrary actuation on any path in the `main` branch. We could do this by using the TF entrypoint path as the path to store the output file(s).

> users need to know _which _ file to read,
It will be the file at the same path as their entrypoint. For example - I have an entrypoint at terraform/gcp-org/x-projects/nebula so the output will go in the same path in the `guardian/outputs` branch (no other files in this branch besides outputs).

> and they have access to all of the other outputs.
This is already true in the current design.

> Teams might have multiple actuations on different sub-folders, which would then clobber.
Not sure how this would occur? Guardian enforces a linear merge process with everything needing to be up to date before merging. Where does the clobbering happen?

> The architecture you describe also has no mechanism for deleting stale output files for a stack that's deleted.
Fair point - this is an existing problem with Guardian though. We don't currently have a way to clean up old statefiles / buckets / entrypoints. We should think more on this generally for Guardian.

> What are the use cases for doing this that can't be satisfied with Terraform remote state?
For X's org - the statefile itself is not accessible to users (only to platform ops SA). As this could contain a lot of sensitive data & I'd like to be able to refactor the state considerably and only maintain that the outputs stay the same (internals vs externals).

>  It would be better if those were hand-selected instead of assuming all outputs need to be published
IMO thats what the terraform outputs are - a hand-selected set of outputs. However, if we wanted to further restrict it we could enforce that each output be marked with a special comment/some indication that it should be shown to users.

> I don't think this is actually going to be as good of a user experience as you think.
I suspect for really complicated configurations you are correct. But I think almost every terraform config in Tidal or X would be ok. 

Overall: I just think that referring back to PR comments to get important configuration is not a great experience.

**Comment by sethvargo** [comment]
> Anecdotally I did this with a pretty complex output for X's org and it seemed pretty readable. I can imagine it would get pretty crazy but maybe putting crazy values in outputs isn't actually helpful anyways?

Readable != parsable though. The use cases described assume some kind of automation is processing the response. For a simple string/number, that might work, but parsing a deeply nested object with variable types will be difficult.

> These values are already all printed out in the PR comment. Anyone with access to the repo can go and see them. Doesn't seem like we are introducing any more risk. Additionally if it is that much of a risk perhaps we should run some kind of secret detection against the output and mask anything that seems like a secret (for PR comments too). If we did that then we could use -json or -raw.

I agree, but the suggestion of creating a branch and making it accessible to automation tooling implies (1) increased access and (2) scope broader than just a platform team. Right now, the threat model is contained to people who already have privileged access. I don't believe we want to give the entire organization access to the infra repos (even if the access is read-only).

> It will be the file at the same path as their entrypoint. For example - I have an entrypoint at terraform/gcp-org/x-projects/nebula so the output will go in the same path in the guardian/outputs branch (no other files in this branch besides outputs).

I think there might be an underlying assumption that the human changing the infra config is the same human that needs access to the output. I do not believe that to be universally true.

> This is already true in the current design.

Not necessarily. Right now, everyone with access to the `infra` repo has access to all outputs. By opening this up to automation, you're expanding the knowledge graph.

> This was assuming an architecture where all the outputs were written to the same file (e.g. outputs.txt). 

> Beyond the initial bootstrapping, what values are needed from Terraform? Most outputs rarely change, so I don't fully understand why we can't just spit out a config file or some markdown that gets copied into the project's README. That's what [we did for EN](https://github.com/google/exposure-notifications-verification-server/blob/main/terraform/main.tf#L166) and it worked super well. If there's a situation where we're updating downstream dependencies from Terraform output values, we're probably holding it incorrectly.

**Comment by dcreey** [comment]
> For a simple string/number, that might work, but parsing a deeply nested object with variable types will be difficult.

Fair enough. I imagine its somewhat self selecting though - since if its hard to parse the user can just update their outputs and flatten/shorten them.

> accessible to automation tooling implies 

I don't actually think automation tooling should have access. I think only users should have read access. The only writer would be the bot running in the same repo (technically that is broader access but seems reasonable?)

> I don't believe we want to give the entire organization access to the infra repos

Yep. I agree. Just people with read rights to the repo.

> I think there might be an underlying assumption that the human changing the infra config is the same human that needs access to the output

I agree with you. Which is why I think putting the info in a PR comment is worse - very difficult to remember where my WIF/SA/bucket details are and how to find them.

> so I don't fully understand why we can't just spit out a config file or some markdown that gets copied into the project's README

Can you elaborate on this? Your link to a file generated by terraform doesn't seem like it will solve the problems (unless this file gets checked in somewhere) since you'd need access to run a tf plan/apply in order to generate the file.

**Comment by sethvargo** [comment]
> Fair enough. I imagine its somewhat self selecting though - since if its hard to parse the user can just update their outputs and flatten/shorten them.

Sure, but again, I think this assumes that the person making the changes to the infra is the same person who needs access to the outputs.

> Can you elaborate on this? Your link to a file generated by terraform doesn't seem like it will solve the problems (unless this file gets checked in somewhere) since you'd need access to run a tf plan/apply in order to generate the file.

I'm still not super clear about the problem you're trying to solve with this proposal. I've heard "abc cli" and "thing I can link to". Here's how I suspect this should work:

- A product team needs a new product environment, they ask the platform ops team to do it, or the propose a PR
- The PR is approved and merged
- Guardian puts the outputs in the comments (project ID, endpoints, etc)
- The product team copies those outputs into their repository (either in a README, script, .env file, etc)
- The product team rarely interacts with outputs again

In the example I linked, Terraform spits out a .env file that includes all the outputs (with some formatting). I'm suggesting we could print those out and product teams can copy-paste them into their product-specific repos.

I think the disconnect is that you're thinking we need continuous access to outputs, and I think of them as a point-in-time exercise that's probably not worth automating.

**Comment by dcreey** [comment]
The more complex (and certainly atypical) use case that I have at X is that that we are attempting to give projects CODEOWNERS on their own directories in our gcp-org repo. In order to do this we need to generate a separate WIF/SA and lock it down guardian to only allow running on a particular directory for a particular WIF/SA. WIF provider is not known until after the apply that creates the WIF. Rather than having a manual step when someone has take the output from apply and add it to a file I'd prefer to just automatically grab the provider details from the output and make it available in an `.env` file for future PR plans/applies. Essentially we could develop a streamlined process for fine-grained auth on an infra repo. Happy to elaborate more.

How would you suggest achieving this? I'd prefer not having the user manually copy data from the PR output (our users know very little about terraform/guardian).

**Comment by sethvargo** [comment]
I'm not sure I follow the use case - can you send me a write-up or link me something in email? Just from the cursory description, it feels like having the apply propose a PR to add that data to the repo would be the best route.
